### DISCUSSION TITLE
Ignore X.509 validations

### DISCUSSION THREAD

**Discussion by oh-mycode** [MatthiasValvekens/pyHanko#84] [Q&A] [answered]
Hi everyone!
I have a pdf document and when I try to verify the digital signature I get the error:
`ValueError: X.509 Name object contains a malformed bidirectional sequence`
As I understood the digital signature has issues with her details related to the signature issuer name.
Is there a way to ignore the error somehow?

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi there!

That depends. The error comes from this bit of logic used in `asn1crypto`'s implementation of `__eq__` for name objects: https://github.com/wbond/asn1crypto/blob/8b6e8035b0805828d35656c38d4dca5d6a145075/asn1crypto/x509.py#L817-L837.

This is not an X.509 validation error, but has to do with an encoding error in the certificate or in the signature's `SignerInfo`---or it's a bug in `asn1crypto`. One of the `Name` objects in the signature data contains a piece of Unicode text that has a malformed bidi sequence (according to `asn1crypto`), which causes problems when trying to compare names _somewhere_.

So, is it possible to suppress this error? Maybe. That depends on where it enters pyHanko's processing---I'll need a complete stack trace to judge that. That said, it's quite likely that pyHanko is hitting this error while trying to _locate_ the signer's certificate, not when trying to validate it. If that's the problem, then there's probably no easy way to solve your issue: failure to match a certificate ID => no signer certificate found => no relevant public key available to validate the signature with.

If you could supply the signed file as well, that'd be even better (see [here](https://github.com/MatthiasValvekens/pyHanko/blob/master/CONTRIBUTING.md#bug-reports) for instructions if you're not comfortable sharing those in public).

---------------------------

TL;DR: This is almost certainly an encoding issue in one of the certificates bundled with the signature,  or in the signature's `SignerInfo`. If not, it's a bug in `asn1crypto`, and there may or may not be an easy workaround.

---------------------------

EDIT: after some back-and-forth over email:

So, the good news is that I found the issue after investigating your files, but the bad news is that it's not going to be easy to work around.


  - First, as I feared, the error occurs during the certificate location phase (i.e. _before_ pyHanko's own certificate validation kicks in), and
the result is that the signer's certificate isn't even found. Since that's a necessary precondition for being able to validate the signature
itself (the certificate is the object that contains the signer's public key), validation would not be able to proceed even if pyHanko would catch and ignore the error.

  - The behaviour in question is not a bug in asn1crypto, but it has to do with the comparison rules for directory names. These use the
stringPrep normalisation profile that's also part of LDAP (RFC 4518, section 1.3). That profile includes restrictions on bidirectional
characters. The gist of it is that in order to prevent rendering ambiguity issues (which could lead to various variants of homograph
attacks), LDAP's stringPrep profile doesn't permit mixing left-to-right and right-to-left characters. See RFC 3454, section 6 for details.

  - The problem with the certificate in your file is that the issuer name is given as "CNקארדקום בעמ", which is a string mixing both
character orientations, and thus can't be processed by LDAP string normalisation.


All things considered, it's pretty much impossible to work around data errors of this type as a user of pyHanko. That said, I do have one
improvement idea that could solve most errors of this type in practice: tell pyHanko to first attempt a byte-for-byte comparison before invoking asn1crypto's implementation of the directory name comparison algorithm.
At least, that should prevent this type of invalid name from causing trouble if it's encoded in _exactly_ the same way in the signature as in
the certificate.

Note: I recently did just that in commit 86400c3.

**Reply by oh-mycode** [reply]
@MatthiasValvekens thank you for your response! 
I have sent to `pyhanko.samples@mvalvekens.be` example of the file that has this issue.
